deployment/ccip/changeset/cs_deploy_chain.go

```diff
@@ -530,43 +561,63 @@ func deployChainContractsSolana(
 	}
 	if initialized {
 		e.Logger.Infow("Router already initialized, skipping initialization", "chain", chain.String())
-		return nil
-	}
-
-	programData, err := solRouterProgramData(e, chain, ccipRouterProgram)
-	if err != nil {
-		return fmt.Errorf("failed to get solana router program data: %w", err)
+	} else {
+		if err := initialzeRouter(e, chain, ccipRouterProgram); err != nil {
+			return err
+		}
 	}
 
-	defaultGasLimit := solBinary.Uint128{Lo: 3000, Hi: 0, Endianness: nil}
-
-	instruction, err := ccip_router.NewInitializeInstruction(
-		chain.Selector,       // chain selector
-		defaultGasLimit,      // default gas limit
-		true,                 // allow out of order execution
-		EnableExecutionAfter, // period to wait before allowing manual execution
-		solana.PublicKey{},
-		GetRouterConfigPDA(ccipRouterProgram),
-		GetRouterStatePDA(ccipRouterProgram),
-		chain.DeployerKey.PublicKey(),
-		solana.SystemProgramID,
-		ccipRouterProgram,
-		programData.Address,
-		GetExternalExecutionConfigPDA(ccipRouterProgram),
-		GetExternalTokenPoolsSignerPDA(ccipRouterProgram),
-	).ValidateAndBuild()
-
-	if err != nil {
-		return fmt.Errorf("failed to build instruction: %w", err)
-	}
-	err = chain.Confirm([]solana.Instruction{instruction})
+	var tokenPoolProgram solana.PublicKey
+	if chainState.SolTokenPool.IsZero() {
+		programID, err := chain.DeployProgram(e.Logger, "token_pool")
+		if err != nil {
+			return fmt.Errorf("failed to deploy program: %w", err)
+		}
+		tv := deployment.NewTypeAndVersion(SolTokenPool, deployment.Version1_0_0)
+		e.Logger.Infow("Deployed contract", "Contract", tv.String(), "addr", programID, "chain", chain.String())
+		tokenPoolProgram = solana.MustPublicKeyFromBase58(programID)
+		err = ab.Save(chain.Selector, programID, tv)
+		if err != nil {
+			return fmt.Errorf("failed to save address: %w", err)
+		}
+	} else {
+		e.Logger.Infow("Using existing token pool", "addr", chainState.SolTokenPool.String())
+		tokenPoolProgram = chainState.SolTokenPool
+	}
+
+	// initialize this last with every address we need
+	if chainState.SolAddressLookupTable.IsZero() {
+		table, err := solCommonUtil.SetupLookupTable(
+			e.GetContext(),
+			chain.Client,
+			*chain.DeployerKey,
+			[]solana.PublicKey{
+				// system
+				solana.SystemProgramID,
+				solana.ComputeBudget,
+				solana.SysVarInstructionsPubkey,
+				// router
+				ccipRouterProgram,
+				GetRouterConfigPDA(ccipRouterProgram),
+				GetRouterStatePDA(ccipRouterProgram),
+				GetSolanaDestChainStatePDA(ccipRouterProgram, chain.Selector),
+				// token pool
+				tokenPoolProgram,
```

Comment:
I assume that as time goes by we'll need more token pools, not just one. As far as I can tell from the code in here, this will be deploying a single `SolTokenPool` and setting its address in the lookup table, right? I guess intially we'll at least have one pool for LINK here, but if we need to support other tokens (and CLL is to deploy & maintain those pools), then it may be worth to include that in this PR... Maybe @RensR can say if we'll need more pools from the beginning or if just a single LINK token pool is enough?

Reply:
Also, per token pool, there are a bunch of account to be used, not just the pool program ID. They include the related token mint and a few others, some of which I guess we can know ahead of time for the tokens we control at least (see [this comment](https://github.com/smartcontractkit/chainlink-ccip/blob/763fd9ec74fac77a9686aaebda86e36a90f678eb/chains/solana/contracts/programs/ccip-router/src/context.rs#L668-L681) and we can chat about which ones we could include here or not)